localstack/aws/protocol/validate.py

```diff
@@ -0,0 +1,173 @@
+
+    error: Error
+
+    def __init__(self, error: Error) -> None:
+        self.error = error
+        super().__init__(self.message)
+
+    @property
+    def reason(self):
+        return self.error.reason
+
+    @property
+    def message(self) -> str:
+        """
+        Returns a default message for the error formatted by BotocoreValidationErrors.
+        :return: the exception message.
+        """
+        return BotocoreValidationErrors()._format_error(self.error)
+
+
+class MissingRequiredField(ParameterValidationError):
+    @property
+    def required_name(self) -> str:
+        return self.error.attributes["required_name"]
+
+
+# TODO: extend subclasses with properties from error arguments as needed. see ValidationErrors._format_error for
+#  which those are.
+
+
+class UnknownField(ParameterValidationError):
+    pass
+
+
+class InvalidType(ParameterValidationError):
+    pass
+
+
+class InvalidRange(ParameterValidationError):
+    pass
+
+
+class InvalidLength(ParameterValidationError):
+    pass
+
+
+class JsonEncodingError(ParameterValidationError):
+    pass
+
+
+class InvalidDocumentType(ParameterValidationError):
+    pass
+
+
+class MoreThanOneInput(ParameterValidationError):
+    pass
+
+
+class EmptyInput(ParameterValidationError):
+    pass
+
+
+class ValidationErrors(BotocoreValidationErrors):
+    def __init__(self, shape, params):
+        super().__init__()
+        self.shape = shape
+        self.params = params
+        self._exceptions: List[ParameterValidationError] = []
+
+    @property
+    def exceptions(self):
+        return self._exceptions
+
+    def raise_first(self):
+        for error in self._exceptions:
+            raise error
+
+    def report(self, name, reason, **kwargs):
+        error = Error(reason, name, kwargs)
+        self._errors.append(error)
+        self._exceptions.append(self.to_exception(error))
+
+    def to_exception(self, error: Error) -> ParameterValidationError:
+        error_type, name, additional = error
+
+        if error_type == "missing required field":
+            return MissingRequiredField(error)
+        elif error_type == "unknown field":
+            return UnknownField(error)
+        elif error_type == "invalid type":
+            return InvalidType(error)
+        elif error_type == "invalid range":
+            return InvalidRange(error)
+        elif error_type == "invalid length":
+            return InvalidLength(error)
+        elif error_type == "unable to encode to json":
+            return JsonEncodingError(error)
+        elif error_type == "invalid type for document":
+            return InvalidDocumentType(error)
+        elif error_type == "more than one input":
+            return MoreThanOneInput(error)
+        elif error_type == "empty input":
+            return EmptyInput(error)
+
+        return ParameterValidationError(error)
+
+
+class ParamValidator(BotocoreParamValidator):
+    def validate(self, params, shape):
```

Comment:
Maybe we could add some type hints here? The shape might not be too obvious for devs who haven't seen any internals of our ASF.

Reply:
yep good point!